# Churn Modelling 

tinyurl.com/ann-classify

## Data Preprocessing

Importing Libraries for Preprocessing

In [1]:
import numpy as np
import matplotlib as plt
import pandas as pd


Step2- Import Dataset

In [2]:
dataset = pd.read_csv('Churn_Modelling.csv')

In [3]:
type(dataset)

pandas.core.frame.DataFrame

In [4]:
dataset

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
1             2    15647311       Hill          608     Spain  Female   41   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
4             5    15737888   Mitchell          850     Spain  Female   43   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0          2       0.00              1          1               1   
1          1   83807.86              1          0               1   
2          8  159660.80              3          1               0   
3          1       0.00              2          0               0   
4          2  125510.82              1          1               1   
...      ...        ...            ...        ...             ...   
9995       5       0.00              2          1               0   
9996      10   57369.61              1          1               1   
9997       7       0.00              1          0               1   
9998       3   75075.31              2          1               0   
9999       4  130142.79              1          1               0   

      EstimatedSalary  Exited  
0           101348.88       1  
1           112542.58       0  
2           113931.57       1  
3            93826.63       0  
4            79084.10       0  
...               ...     ...  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0  

[10000 rows x 14 columns]

In [5]:
x=dataset.iloc[:,3:13]

Credit scoring is a statistical analysis performed by lenders and financial institutions to access a person's creditworthiness. Credit scoring is used by lenders to help decide on whether to extend or deny credit. A person's credit score is a number between 300 and 850, 850 being the highest credit rating possible. 

In [6]:
x

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0             619    France  Female   42       2       0.00              1   
1             608     Spain  Female   41       1   83807.86              1   
2             502    France  Female   42       8  159660.80              3   
3             699    France  Female   39       1       0.00              2   
4             850     Spain  Female   43       2  125510.82              1   
5             645     Spain    Male   44       8  113755.78              2   
6             822    France    Male   50       7       0.00              2   
7             376   Germany  Female   29       4  115046.74              4   
8             501    France    Male   44       4  142051.07              2   
9             684    France    Male   27       2  134603.88              1   
10            528    France    Male   31       6  102016.72              2   
11            497     Spain    Male   24       3       0.00              2   
12            476    France  Female   34      10       0.00              2   
13            549    France  Female   25       5       0.00              2   
14            635     Spain  Female   35       7       0.00              2   
15            616   Germany    Male   45       3  143129.41              2   
16            653   Germany    Male   58       1  132602.88              1   
17            549     Spain  Female   24       9       0.00              2   
18            587     Spain    Male   45       6       0.00              1   
19            726    France  Female   24       6       0.00              2   
20            732    France    Male   41       8       0.00              2   
21            636     Spain  Female   32       8       0.00              2   
22            510     Spain  Female   38       4       0.00              1   
23            669    France    Male   46       3       0.00              2   
24            846    France  Female   38       5       0.00              1   
25            577    France    Male   25       3       0.00              2   
26            756   Germany    Male   36       2  136815.64              1   
27            571    France    Male   44       9       0.00              2   
28            574   Germany  Female   43       3  141349.43              1   
29            411    France    Male   29       0   59697.17              2   
...           ...       ...     ...  ...     ...        ...            ...   
9970          518    France    Male   42       7  151027.05              2   
9971          833    France  Female   34       3  144751.81              1   
9972          758    France    Male   26       4  155739.76              1   
9973          611    France    Male   27       7       0.00              2   
9974          583    France    Male   33       7  122531.86              1   
9975          610   Germany    Male   50       1  113957.01              2   
9976          637    France  Female   33       7  103377.81              1   
9977          683    France  Female   32       9       0.00              2   
9978          774    France    Male   40       9   93017.47              2   
9979          677    France  Female   58       1   90022.85              1   
9980          741     Spain    Male   35       6   74371.49              1   
9981          498   Germany    Male   42       3  152039.70              1   
9982          655   Germany  Female   46       7  137145.12              1   
9983          613    France    Male   40       4       0.00              1   
9984          602   Germany    Male   35       7   90602.42              2   
9985          659    France    Male   36       6  123841.49              2   
9986          673   Germany    Male   47       1  183579.54              2   
9987          606     Spain    Male   30       8  180307.73              2   
9988          775    France    Male   30       4       0.00              2   
9989          841     Spain 

In [7]:
x=dataset.iloc[:,3:13].values

In [8]:
x

array([[619, 'France', 'Female', ..., 1, 1, 101348.88],
       [608, 'Spain', 'Female', ..., 0, 1, 112542.58],
       [502, 'France', 'Female', ..., 1, 0, 113931.57],
       ...,
       [709, 'France', 'Female', ..., 0, 1, 42085.58],
       [772, 'Germany', 'Male', ..., 1, 0, 92888.52],
       [792, 'France', 'Female', ..., 1, 0, 38190.78]], dtype=object)

In [9]:
y=dataset.iloc[:,13].values

In [10]:
y

array([1, 0, 1, ..., 1, 1, 0], dtype=int64)

In [11]:
dataset.isnull()

RowNumber  CustomerId  Surname  CreditScore  Geography  Gender    Age  \
0         False       False    False        False      False   False  False   
1         False       False    False        False      False   False  False   
2         False       False    False        False      False   False  False   
3         False       False    False        False      False   False  False   
4         False       False    False        False      False   False  False   
5         False       False    False        False      False   False  False   
6         False       False    False        False      False   False  False   
7         False       False    False        False      False   False  False   
8         False       False    False        False      False   False  False   
9         False       False    False        False      False   False  False   
10        False       False    False        False      False   False  False   
11        False       False    False        False      False   False  False   
12        False       False    False        False      False   False  False   
13        False       False    False        False      False   False  False   
14        False       False    False        False      False   False  False   
15        False       False    False        False      False   False  False   
16        False       False    False        False      False   False  False   
17        False       False    False        False      False   False  False   
18        False       False    False        False      False   False  False   
19        False       False    False        False      False   False  False   
20        False       False    False        False      False   False  False   
21        False       False    False        False      False   False  False   
22        False       False    False        False      False   False  False   
23        False       False    False        False      False   False  False   
24        False       False    False        False      False   False  False   
25        False       False    False        False      False   False  False   
26        False       False    False        False      False   False  False   
27        False       False    False        False      False   False  False   
28        False       False    False        False      False   False  False   
29        False       False    False        False      False   False  False   
...         ...         ...      ...          ...        ...     ...    ...   
9970      False       False    False        False      False   False  False   
9971      False       False    False        False      False   False  False   
9972      False       False    False        False      False   False  False   
9973      False       False    False        False      False   False  False   
9974      False       False    False        False      False   False  False   
9975      False       False    False        False      False   False  False   
9976      False       False    False        False      False   False  False   
9977      False       False    False        False      False   False  False   
9978      False       False    False        False      False   False  False   
9979      False       False    False        False      False   False  False   
9980      False       False    False        False      False   False  False   
9981      False       False    False        False      False   False  False   
9982      False       False    False        False      False   False  False   
9983      False       False    False        False      False   False  False   
9984      False       False    False        False      False   False  False   
9985      False       False    False        False      False   False  False   
9986      False       False    False        False      False   False  False   
9987      False       False    False        False      False   False  False   
9988      False       False    False        False      F

# Label Encoding 

<img src="https://ekababisong.org/assets/seminar_IEEE/LabelEncoder.png" width=1000 height=10500 />

Example 2 

<img src="https://raw.githubusercontent.com/WillKoehrsen/Machine-Learning-Projects/master/label_encoding.png" width=500 height=500 />

Step5-Encode categorical Text Variables

In [12]:
from sklearn.preprocessing import LabelEncoder
lb=LabelEncoder()
lb2=LabelEncoder()

In [13]:
x[:,1]

array(['France', 'Spain', 'France', ..., 'France', 'Germany', 'France'],
      dtype=object)

In [14]:
x[:,1]=lb.fit_transform(x[:,1])

In [15]:
x

array([[619, 0, 'Female', ..., 1, 1, 101348.88],
       [608, 2, 'Female', ..., 0, 1, 112542.58],
       [502, 0, 'Female', ..., 1, 0, 113931.57],
       ...,
       [709, 0, 'Female', ..., 0, 1, 42085.58],
       [772, 1, 'Male', ..., 1, 0, 92888.52],
       [792, 0, 'Female', ..., 1, 0, 38190.78]], dtype=object)

In [16]:
x[:,2]=lb2.fit_transform(x[:,2])#if you have more than one more that one category of categorical variable then we have to do label encoding all at once

In [17]:
x

array([[619, 0, 0, ..., 1, 1, 101348.88],
       [608, 2, 0, ..., 0, 1, 112542.58],
       [502, 0, 0, ..., 1, 0, 113931.57],
       ...,
       [709, 0, 0, ..., 0, 1, 42085.58],
       [772, 1, 1, ..., 1, 0, 92888.52],
       [792, 0, 0, ..., 1, 0, 38190.78]], dtype=object)

# one-hot encoding

<img src="https://i.imgur.com/mtimFxh.png" width=500 height=500 />


In [18]:
from sklearn.preprocessing import OneHotEncoder
oh=OneHotEncoder(categorical_features=[1])
x=oh.fit_transform(x).toarray()  

In [19]:
x

array([[1.0000000e+00, 0.0000000e+00, 0.0000000e+00, ..., 1.0000000e+00,
        1.0000000e+00, 1.0134888e+05],
       [0.0000000e+00, 0.0000000e+00, 1.0000000e+00, ..., 0.0000000e+00,
        1.0000000e+00, 1.1254258e+05],
       [1.0000000e+00, 0.0000000e+00, 0.0000000e+00, ..., 1.0000000e+00,
        0.0000000e+00, 1.1393157e+05],
       ...,
       [1.0000000e+00, 0.0000000e+00, 0.0000000e+00, ..., 0.0000000e+00,
        1.0000000e+00, 4.2085580e+04],
       [0.0000000e+00, 1.0000000e+00, 0.0000000e+00, ..., 1.0000000e+00,
        0.0000000e+00, 9.2888520e+04],
       [1.0000000e+00, 0.0000000e+00, 0.0000000e+00, ..., 1.0000000e+00,
        0.0000000e+00, 3.8190780e+04]])

In [20]:
x=x[:,1:]

In [21]:
x[1,:]

array([0.0000000e+00, 1.0000000e+00, 6.0800000e+02, 0.0000000e+00,
       4.1000000e+01, 1.0000000e+00, 8.3807860e+04, 1.0000000e+00,
       0.0000000e+00, 1.0000000e+00, 1.1254258e+05])

In [22]:
from sklearn.cross_validation import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.3,random_state=0)

C:\Users\HP\Anaconda3\lib\site-packages\sklearn\cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


# Standard Scaling 

<img src="https://i.stack.imgur.com/PZgJ2.pngg" width=500 height=500 />


In [23]:
from sklearn.preprocessing import StandardScaler
sc= StandardScaler()

x_train =sc.fit_transform(x_train)
x_test = sc.transform(x_test)

In [24]:
import keras
from keras.models import Sequential
from keras.layers import Dense

C:\Users\HP\Anaconda3\lib\site-packages\h5py\__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


In [25]:
keras.__version__

'2.2.4'

In [26]:
model=Sequential()

In [27]:
x_train.shape

(7000, 11)

In [28]:
model.add(Dense(input_dim=11,init="random_uniform",activation='relu',output_dim=15))#input layer

Instructions for updating:
Colocations handled automatically by placer.


C:\Users\HP\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(input_dim=11, activation="relu", units=15, kernel_initializer="random_uniform")`
  """Entry point for launching an IPython kernel.


In [29]:
model.add(Dense(output_dim=8,init="random_uniform",activation='relu'))#hidden layer

C:\Users\HP\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", units=8, kernel_initializer="random_uniform")`
  """Entry point for launching an IPython kernel.


In [30]:
model.add(Dense(output_dim=1,init='random_uniform',activation='sigmoid'))#output layer

C:\Users\HP\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="sigmoid", units=1, kernel_initializer="random_uniform")`
  """Entry point for launching an IPython kernel.


# since it is a binary classification WE use binary cross entropy.if it is categorical then we ue categorical entropy

In [31]:
model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])#adam=batch gradent descent

In [32]:
model.fit(x_train,y_train,epochs=50,batch_size=32)#epochs no of iteration 

Instructions for updating:
Use tf.cast instead.
Epoch 1/50
7000/7000 [==============================] - 1s 139us/step - loss: 0.5714 - acc: 0.7951
Epoch 2/50
7000/7000 [==============================] - 0s 69us/step - loss: 0.4327 - acc: 0.7977
Epoch 3/50
7000/7000 [==============================] - 0s 63us/step - loss: 0.4261 - acc: 0.7977
Epoch 4/50
7000/7000 [==============================] - 1s 74us/step - loss: 0.4218 - acc: 0.7977
Epoch 5/50
7000/7000 [==============================] - 0s 63us/step - loss: 0.4174 - acc: 0.8166
Epoch 6/50
7000/7000 [==============================] - 0s 59us/step - loss: 0.4136 - acc: 0.8280
Epoch 7/50
7000/7000 [==============================] - 0s 60us/step - loss: 0.4102 - acc: 0.8310
Epoch 8/50
7000/7000 [==============================] - 0s 61us/step - loss: 0.4072 - acc: 0.8331
Epoch 9/50
7000/7000 [==============================] - 0s 62us/step - loss: 0.4046 - acc: 0.8339
Epoch 10/50
7000/7000 [==============================] - 0s 61us/step

In [33]:
y_pred = model.predict(x_test)

In [51]:
y_pred=(y_pred>0.5)

In [58]:
y_pred

array([[False],
       [False],
       [False],
       ...,
       [False],
       [False],
       [ True]])

In [52]:
y_test

array([0, 1, 0, ..., 0, 0, 1], dtype=int64)

In [53]:
y_p=model.predict(sc.transform(np.array([[0,1,100,1,25,8,115000,24,1,1,100000]])))

C:\Users\HP\Anaconda3\lib\site-packages\sklearn\utils\validation.py:475: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)


In [54]:
y_p

array([[1.]], dtype=float32)

In [55]:
y_p=(y_p>0.5)

In [56]:
y_p

array([[ True]])

In [57]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_pred)

0.855